lib/phoenix/tracker.ex

```diff
@@ -62,13 +62,48 @@ defmodule Phoenix.Tracker do
   An optional `handle_info/2` callback may also be invoked to handle
   application specific messages within your tracker.
 
-  ## Special Considerations
+  ## Stability and Performance Considerations
 
   Operations within `handle_diff/2` happen *in the tracker server's context*.
   Therefore, blocking operations should be avoided when possible, and offloaded
   to a supervised task when required. Also, a crash in the `handle_diff/2` will
   crash the tracker server, so operations that may crash the server should be
   offloaded with a `Task.Supervisor` spawned process.
+
+  ## Application Shutdown
+  
+  A tracker does not automatically replicate its state across the cluster as it
+  shuts down. This means that your supervision tree shuts down normally - as it
+  does when you call `System.stop()` or the BEAM receives a `SIGTERM` - any
+  presences that the local tracker instance has will continue to be seen as
+  present by other trackers in the cluster until the `:down_period` for the
+  instance has passed.
+
+  To update the cluster more quickly during a normal shutdown, you can add a
+  GenServer to your supervision tree after the tracker instance (to be started
+  after it and shut down before it) which calls `graceful_permdown/1` on the
+  tracker as it shuts down.
+
+      defmodule MyApp.TrackerShutdownHandler do
+        use GenServer
+      
+        def start_link(opts \\\\ []) do
+          GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
+        end
+      
+        @impl GenServer
+        def init(init_arg) do
+          # So that terminate/2 will be called during graceful shutdown
+          Process.flag(:trap_exit, true)
+          {:ok, init_arg}
+        end
+      
+        @impl GenServer
+        def terminate(_reason, _state) do
+          Phoenix.Tracker.graceful_permdown(MyAppWeb.Tracker)
+          :ok
+        end
+      end
```

Comment:
The problem with this approach is that it does not distinguish between a clean terminate and a non-clean one. Maybe we should check the reason? But I am not sure if the supervisor will pass the relevant information either.

Reply:
Are you referring to the ignored `_reason`? My thinking was that since this GenServer doesn't do anything except wait for termination, it should only terminate when the supervision tree is shutting down, at which point the Tracker is going to be the next thing to go down. When might something go wrong here?

Reply:
If the BEAM crashes, this `terminate` obviously won't run, but you'd be no worse off than without it; you'd be back to having the other nodes wait the `:down_period`.

Reply:
> it should only terminate when the supervision tree is shutting down

Yes, but a supervision tree may also shut down and be restarted. Or this is being shutdown because the restart strategy is `:rest_for_one` or `:one_for_all`. So this is assuming this is in the root supervision tree AND the strategy is `:one_for_one`. We should probably make these assumptions clearer?

Reply:
I also wonder then why we don't add this as an option to the tracker itself:

    perm_down_on_shutdown: ...

But maybe there are other implications there.

Reply:
I also wonder if we should add a `{Task.Shutdown, fn -> ... end}` to Elixir that makes it easier to run something only on shutdown.

Reply:
> Yes, but a supervision tree may also shut down and be restarted.
Even if that's the case, if the tracker is about to shut down, it's about to lose its state, right? Anything that it has tracked, it will no longer know about until they call `.track` again after the restart. If it's state is about to be lost, we can either have the other nodes find out about that or be out of sync?

As to your ideas for building this in somehow, I would love that. I was very surprised that this didn't get handled automatically, so part of the reason for opening this PR was to confirm/deny that this was the right approach.

Reply:
> Even if that's the case, if the tracker is about to shut down, it's about to lose its state, right?

Yes, but in Phoenix.Presence cases, the clients will reconnect, and not issuing a perm down means that clients don't immediately get a bunch of notifications that thousands or dozens of thousands of users left. They will see new metadata joining and then leaving, but with a time between them. Instead of leaves and joins.

In any case, I think we can have a `permdown_on_shutdown` option which can be set to true and bake this in. :)

Reply:
Do you want some modified version of this documentation in the meantime? Do you want a PR for `perm_on_shutdown`? (If so, I could use some guidance on where to implement that.)

Reply:
>  Do you want a PR for perm_on_shutdown? (If so, I could use some guidance on where to implement that.)

Yes, please! `Phoenix.Tracker` is a supervisor, you need to add a child similar to the one you documented here. :)